src/java.base/share/classes/javax/crypto/KDFSpi.java

```diff
@@ -0,0 +1,112 @@
+/**
+ * This class defines the <i>Service Provider Interface</i> (<b>SPI</b>) for the
+ * <code>KeyDerivation</code> class.
+ * <p>
+ * All the abstract methods in this class must be implemented by each cryptographic service provider
+ * who wishes to supply the implementation of a particular key derivation algorithm.
+ */
+public abstract class KDFSpi {
+
+    /**
+     * provides access to the KDF alg params for implementers of the Spi
+     */
+    protected final AlgorithmParameterSpec algorithmParameterSpec;
+
+    /**
+     * The sole constructor.
+     * <p>
+     * TODO: additional description
+     *
+     * @param algParameterSpec
+     *     the initialization parameters for the KDF algorithm (may be {@code null})
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the initialization parameters are inappropriate for this {@code KDFSpi}
+     */
+    protected KDFSpi(AlgorithmParameterSpec algParameterSpec)
+        throws InvalidAlgorithmParameterException {
+        this.algorithmParameterSpec = algParameterSpec;
+    }
+
+
+    /**
+     * Derive a key, returned as a {@code Key}.
+     * <p>
+     * TODO: additional description
+     *
+     * @param alg
+     *     the algorithm of the resultant key object
+     * @param kdfParameterSpec
+     *     derivation parameters
+     *
+     * @return a {@code SecretKey} object corresponding to a key built from the KDF output and according
+     * to the derivation parameters.
+     *
+     * @throws InvalidParameterSpecException
+     *     if the information contained within the {@code DerivationParameterSpec} is invalid or
+     *     incorrect for the type of key to be derived, or specifies a type of output that is not a
+     *     key (e.g. raw data)
+     * @throws IllegalStateException
+     *     if the key derivation implementation cannot support additional calls to {@code deriveKey}
+     *     or if all {@code DerivationParameterSpec} objects provided at initialization have been
+     *     processed.
+     */
+    protected abstract SecretKey engineDeriveKey(String alg, KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException;
+
+    /**
+     * Obtain raw data from a key derivation function.
+     * <p>
+     * TODO: additional description
+     *
+     * @param kdfParameterSpec
+     *     derivation parameters
+     *
+     * @return a byte array whose length matches the length field in the processed
+     * {@code DerivationParameterSpec} and containing the next bytes of output from the key
+     * derivation function.
+     *
+     * @throws InvalidParameterSpecException
+     *     if the {@code DerivationParameterSpec} being applied to this method is of a type other
+     *     than "data".
+     * @throws IllegalStateException
+     *     if the key derivation implementation cannot support additional calls to
+     *     {@code deriveData } or if all {@code DerivationParameterSpec} objects have been
+     *     processed.
```

Comment:
What kind of exception will be thrown if the derived key of this KDF is not extractable.

Reply:
Sounds like we may need a new exception class, like `DerivationException`.